apps/gateway/test/champions_test.exs

```diff
@@ -0,0 +1,286 @@
+      muflus = GameBackend.Units.Characters.get_character_by_name("Muflus")
+      {:ok, user} = Users.register("LevelUpUser")
+      Currencies.add_currency_by_name!(user.id, "Gold", 9999)
+
+      {:ok, unit} =
+        GameBackend.Units.insert_unit(%{
+          user_id: user.id,
+          level: 19,
+          tier: 1,
+          rank: 2,
+          selected: false,
+          character_id: muflus.id
+        })
+
+      gold = Currencies.get_amount_of_currency_by_name(user.id, "Gold")
+      level = unit.level
+
+      #### Level up
+      [%CurrencyCost{currency_id: _gold_id, amount: level_up_cost}] = Units.calculate_level_up_cost(unit)
+
+      :ok = SocketTester.level_up_unit(socket_tester, user.id, unit.id)
+      fetch_last_message(socket_tester)
+
+      assert_receive %WebSocketResponse{
+        response_type: {:unit_and_currencies, %UnitAndCurrencies{}}
+      }
+
+      fetch_last_message(socket_tester)
+
+      %WebSocketResponse{
+        response_type: {:unit_and_currencies, %UnitAndCurrencies{unit: unit, user_currency: [user_currency]}}
+      } = get_last_message()
+
+      assert unit.level == level + 1
+      assert user_currency.currency.name == "Gold"
+      assert user_currency.amount == gold - level_up_cost
+
+      # Cannot level up because unit is level 20 with tier 1
+      assert unit.level == 20
+      assert unit.tier == 1
+      :ok = SocketTester.level_up_unit(socket_tester, user.id, unit.id)
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:error, %Error{reason: "cant_level_up"}}}
+
+      #### Tier up
+
+      user_gold = Currencies.get_amount_of_currency_by_name(user.id, "Gold")
+      user_gems = Currencies.get_amount_of_currency_by_name(user.id, "Gems")
+
+      [%CurrencyCost{currency_id: _gold_id, amount: gold_cost}, %CurrencyCost{currency_id: _gems_id, amount: gems_cost}] =
+        Units.calculate_tier_up_cost(unit)
+
+      :ok = SocketTester.tier_up_unit(socket_tester, user.id, unit.id)
+      fetch_last_message(socket_tester)
+
+      assert_receive %WebSocketResponse{
+        response_type: {:unit_and_currencies, %UnitAndCurrencies{}}
+      }
+
+      fetch_last_message(socket_tester)
+
+      %WebSocketResponse{
+        response_type: {:unit_and_currencies, %UnitAndCurrencies{unit: unit, user_currency: user_currencies}}
+      } = get_last_message()
+
+      user_currencies =
+        Enum.into(user_currencies, %{}, fn %UserCurrency{
+                                             currency: %Currency{name: name},
+                                             amount: amount
+                                           } ->
+          {name, amount}
+        end)
+
+      assert unit.level == 20
+      assert unit.tier == 2
+      assert user_currencies["Gold"] == user_gold - gold_cost
+      assert user_currencies["Gems"] == user_gems - gems_cost
+
+      # TODO: Check that we can now level up
```

Comment:
Is this todo for another pr?

Reply:
I think I'd left these two as TO-DO so as to not make the PR too long. What do you think, should I implement it here?

Reply:
hmmmm since it's not that big of a chance i would add it here, if we would tests a new functionality yeah we should add it in another pr so this one doesn't scale, but this is just repeating the previous code right?